### PR TITLE
Add `__str__` and `__repr__` for components and pipelines

### LINKED ISSUE #474: Add `__str__`+`__repr__` for `ComponentBase`+`PipelineBase`
- __str__
    - Printing a pipeline or component instance should quickly identify which pipeline or component it is.
    - Components
        - Use ComponentBase.name
     - Pipelines
        - Use PipelineBase.summary
- __repr__
    - Calling repr() on a pipeline or component should produce useful output for debugging.
    - Components
        - print Component and parameters like python expression
        - ComponentBase(param1=1, param2=2)
    - Pipelines
            - print python expression with parameters dict and objective
            - LogisticRegressionPipeline(parameters={...}, objective='precision
                - parameters = {'LGC': {'C':1.5}, 'SI':{'Impute Strategy': 'mean}}

```
// component
def __str__(self):
        return self.name 

// pipeline
def __str__(self):
        return self.name

// component
def __repr__(self):
        rpr = f"{(type(self).__name__)}("
        for parameter, value in self.parameters.items():
             rpr = rpr + f"{parameter}={value},"
        rpr = + ')'
        return rpr
    
// pipeline
def __repr__(self):
        rpr = f"{(type(self).__name__)}(parameters={"
        for component, parameters in self.parameters.items():
            rpr = rpr + f"'\n{component}':{"
            for parameter, value in parameters:
                rpr = rpr + f"\n{parameter}={value},"
            rpr = rpr + "},"
         rpr = + ')'
         return rpr
```

### PR DESCRIPTION
**Pull request by eccabay**
Closes #474

### CODE REVIEW

File: evalml/pipelines/components/component_base.py
Comment (dsherry) [inline evalml/pipelines/components/component_base.py:168]: @eccabay good call looking into `float('inf')`!! Another one is `float('nan')`. You've hit on [what appears to be a funny bug and/or design flaw with python's `float` type](https://bugs.python.org/issue22951): `repr(float('inf'))` comes back as `'inf'`, which... is just messed up! 🤔   Same for `float('nan')`. I think they did this so that `float(repr(float('inf')))` is equivalent to `float('inf')`, and same for nan.

The same is true for `np.inf` and `np.nan`, by the way:
```python
In [6]: parameters = {'int': 42, 'string': 'string', 'float': 3.14159,
                      'inf': float('inf'), 'np inf': np.inf, 'np nan': np.nan}

In [7]: repr(parameters)
Out[7]: "{'int': 42, 'string': 'string', 'float': 3.14159, 'inf': inf, 'np inf': inf, 'np nan': nan}"
```

---
So, what do we do? It would be really nice if we could have what comes out of `repr(component)` be something we could paste into a terminal in order to get an identical instance.

I like the approach you're taking. I played with it for a bit and here's what I got:
```python
# define this in gen_utils so we can use it for the pipeline repr too
def safe_repr(value):
    if isinstance(value, float):
        if pd.isna(value):
            return 'np.nan'
        if np.isinf(value):
            return f'float({repr(value)})'
    return repr(value)

# then the component repr:
def __repr__(self):
    parameters_repr = ', '.join([f'{key}={safe_repr(value)}' for key, value in self.parameters.items()])
    return f'{self.name}({parameters_repr})'
```

Here's a test I did of the `safe_repr`, which shows that it outputs the string we need in order to avoid some ugly bugs:
```
In [1]: [safe_repr(el) for el in ['string', float('nan'), float('-inf'), float('inf'), np.nan, -np.inf, np.inf]]
Out[1]:
["'string'", 'np.nan', 'float(-inf)', 'float(inf)', 'np.nan', 'float(-inf)', 'float(inf)']
```

What do you think of that? Weird stuff!

File: evalml/pipelines/pipeline_base.py
Comment (dsherry) [inline evalml/pipelines/pipeline_base.py:507]: Same comment as for component repr, let's flatten this out, and great thinking trying nan/inf types!

File: evalml/tests/pipeline_tests/test_pipelines.py
Comment (dsherry) [inline evalml/tests/pipeline_tests/test_pipelines.py:1317]: This test is pretty cool! Great thinking enforcing this invariant on `repr`

Two things:
1. RE our discussion of nan types at the top, it would be ideal if we had a similar test to this one which checks all those edge cases. Example: define a mock component which has a few input parameters, and pass in `float('nan')`/`np.nan` and the native/np infs.

2. There are [security concerns with using python's `eval`](https://stackoverflow.com/questions/3513292/python-make-eval-safe). It turns out if you can compromise what goes into `eval` you can do a lot of bad stuff. To avoid this entirely, my recommendation is to avoid using `eval` and just check the string output matches what we expect. I know, this is test code so it feels kinda silly, lol--[related reading](https://rules.sonarsource.com/python/type/Security%20Hotspot/RSPEC-1523) if you're interested.

**Review by dsherry — CHANGES_REQUESTED** [review]
@eccabay great work on this! There's some seriously tricky stuff going on here for a feature which looks so simple from the outside. It's really cool we're doing this :)

I left a suggested change on the impl along with some discussion. I also left a couple items on testing. Should be all set after that IMO.

File: evalml/pipelines/components/component_base.py
Comment (dsherry) [inline evalml/pipelines/components/component_base.py:169]: Please delete this dead code

File: evalml/utils/gen_utils.py
Comment (dsherry) [inline evalml/utils/gen_utils.py:238]: Please add a docstring!

File: evalml/tests/pipeline_tests/test_pipelines.py
Comment (dsherry) [inline evalml/tests/pipeline_tests/test_pipelines.py:1323]: Looks great! I left a comment on the impl about how to get rid of the double curly braces here.

File: evalml/pipelines/pipeline_base.py
Comment (dsherry) [inline evalml/pipelines/pipeline_base.py:501]: I believe [you want double curly braces instead of triple](https://stackoverflow.com/questions/5466451/how-can-i-print-literal-curly-brace-characters-in-python-string-and-also-use-fo) here.

**Review by dsherry — APPROVED** [review]
Sweet!! Approved pending a couple small things:
* Remove extra curly braces from pipeline repr
* Add docstring for `safe_repr`

File: evalml/pipelines/pipeline_base.py
Comment (eccabay) [reply evalml/pipelines/pipeline_base.py:501]: Since the parameters is a dictionary and contains an expression, I do need all three! The first two are for literal curly braces and the third is for the formatting.

Comment (dsherry) [reply evalml/pipelines/pipeline_base.py:501]: Hm interesting, but then the output (taken from the unit test) ends up as
```python
MockPipeline(parameters={{'Imputer':{{'categorical_impute_strategy': 'most_frequent', 'numeric_impute_strategy': 'mean', 'categorical_fill_value': None, 'numeric_fill_value': None}}, '{final_estimator}':{{'n_estimators': 100, 'max_depth': 6, 'n_jobs': -1}},}})
```

But ideally I think it should be
```python
MockPipeline(parameters={'Imputer': {'categorical_impute_strategy': 'most_frequent', 'numeric_impute_strategy': 'mean', 'categorical_fill_value': None, 'numeric_fill_value': None}, 'final_estimator': {'n_estimators': 100, 'max_depth': 6, 'n_jobs': -1}})
```

Do you agree? The second output could be evaluated in the python repl and turned into an object. I think the first would fail.

File: evalml/tests/pipeline_tests/test_pipelines.py
Comment (dsherry) [inline evalml/tests/pipeline_tests/test_pipelines.py:1307]: @eccabay I just noticed: can we replace `'{final_estimator}':` with `'final_estimator':`?

Comment (dsherry) [reply evalml/tests/pipeline_tests/test_pipelines.py:1307]: Doesn't block merge though IMO! Just a detail we should probably chase down

File: evalml/pipelines/pipeline_base.py
Comment (dsherry) [reply evalml/pipelines/pipeline_base.py:501]: @eccabay but, I think its fine to merge this and we can circle back. It adds value even if there's a couple details we may wanna discuss further :)

Comment (eccabay) [reply evalml/pipelines/pipeline_base.py:501]: The double curly braces in the unit tests are once again for f-string formatting! If you actually print the repr, you get the second code block you pasted there, and calling eval on the expected_repr produces the correct object.

Comment (dsherry) [reply evalml/pipelines/pipeline_base.py:501]: @eccabay hmm could you paste an example? I don't follow yet. I thought that `__repr__` returns a string, which requires no further formatting, and that's the end of it 😂  In other words, I thought we should output the second snippet, because that can be pasted into the python REPL and evaluated, but we're currently outputting the first.

Again, not blocking merge, I'm just trying to make sure we all agree on desired behavior.

File: evalml/tests/pipeline_tests/test_pipelines.py
Comment (eccabay) [reply evalml/tests/pipeline_tests/test_pipelines.py:1307]: Unfortunately the way these tests are written, we can't. The string `__repr__` prints out uses the name of the estimator, so checking for string equality will fail. If I replace `{final_estimator}` with `final_estimator` and try `assert eval(repr(pipeline)) == eval(expected_repr)`, that will pass, but that will reintroduce `eval` to the code.

The alternative of having `__repr__` output `final_estimator` instead of the name of said estimator feels unnecessarily clunky.

Comment (dsherry) [reply evalml/tests/pipeline_tests/test_pipelines.py:1307]: Oh! My bad, I didn't notice this was a format string until just now. Got it, makes sense. Thanks

Comment (eccabay) [reply evalml/tests/pipeline_tests/test_pipelines.py:1307]: Wonderful, glad we're on the same page :sweat_smile:. Does this also clear up our other discussion?

File: evalml/pipelines/pipeline_base.py
Comment (dsherry) [reply evalml/pipelines/pipeline_base.py:501]: @eccabay ok, I think I understand now. So the string I copied was a format string, which is why you need the double curlys there.

I just did a test locally and everything looks great

Screen Shot 2020-09-29 at 10 40 39 AM

Realizing the unit tests were using a format string was the key point I was missing, lol. Thanks! 😆

**Pull request merged**
